platforms/cpuinfo_linux.go

```diff
@@ -0,0 +1,88 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package platforms
+
+import (
+	"bytes"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"golang.org/x/sys/unix"
+)
+
+// getMachineArch retrieves the machine architecture through system call
+// same result from uname -m running the terminal
+// This function is favored to the previous method of getting variant information from /proc/cpuinfo
+// because /proc/cpuinfo is not accurate in user-mode emulation environment
+func getMachineArch() (string, error) {
+	var uname unix.Utsname
+	err := unix.Uname(&uname)
+	if err != nil {
+		return "", err
+	}
+
+	arch := string(uname.Machine[:bytes.IndexByte(uname.Machine[:], 0)])
+
+	return arch, nil
+}
+
+// getCPUVariant returns cpu variant for ARM
+// return empty string in case of error
+func getCPUVariant() string {
+
+	if !isArmArch(runtime.GOARCH) {
+		log.L.Errorf("getCPUVariant for ARCH %s: %v", runtime.GOARCH, errdefs.ErrNotImplemented)
+	}
```

Comment:
How about having this isArmArch check in either the caller or the function itself, not both?

Reply:
I added a check inside the function just in case it is accidentally invoked by Linux/amd platforms. I can remove the check inside the function.